pkg/crd/markers/validation.go

```diff
@@ -284,3 +294,55 @@ func (m Nullable) ApplyToSchema(schema *v1beta1.JSONSchemaProps) error {
 	schema.Nullable = true
 	return nil
 }
+
+// Defaults are only valid CRDs created with the v1 API
+func (m Default) ApplyToSchema(schema *v1beta1.JSONSchemaProps) error {
+	// Convert versioned schema to unversioned form
+	marshalledSchema, err := json.Marshal(schema)
+	if err != nil {
+		return err
+	}
+	unversionedSchema := &apiext.JSONSchemaProps{}
+	err = json.Unmarshal(marshalledSchema, unversionedSchema)
+	if err != nil {
+		return err
+	}
+
+	// Initialize validator for schema
+	crValidation := &apiext.CustomResourceValidation{
+		OpenAPIV3Schema: unversionedSchema,
+	}
+	validator, _, err := validation.NewSchemaValidator(crValidation)
+	if err != nil {
+		return err
+	}
+
+	value := string(m)
+
+	// Quote strings for json compatibility.
+	if schema.Type == "string" {
+		value = fmt.Sprintf("%q", value)
+	}
+
+	// Unmarshal the supplied default to an empty interface
+	var obj interface{}
+	err = json.Unmarshal([]byte(value), &obj)
+	if err != nil {
+		return err
+	}
+
+	// Validate the default with the field's schema.
+	result := validator.Validate(obj)
+	if len(result.Errors) > 0 {
+		return fmt.Errorf("default fails validation: %v", result.Errors)
+	}
```

Comment:
note that this is by far not complete. This lacks 
- the apiextensions/v1 validation
- a check for structurality of the provided schema
- a check that we actually have a v1 CRD
- validation that the value is pruned
- logic to cope with `x-kubernetes-preserve-unknown-fields`
- logic to handle defaults under ObjectMeta correctly.

Reply:
We don't want to and cannot do this validation here.

Reply:
I'm going to treat this as an educational opportunity. :)  

> note that this is by far not complete. This lacks
> 
> * the apiextensions/v1 validation

What is this and how is it relevant to default validation?

> * a check for structurality of the provided schema

This method is validating a default value, not a schema. Am I missing something?

> * a check that we actually have a v1 CRD

As previously discussed, the current architecture of schema generation will likely require removal of generated defaults via post-processing.

> * validation that the value is pruned

Is this something I should just ignore then?

> * logic to cope with `x-kubernetes-preserve-unknown-fields`

Hmm, I was optimistically hoping that the validation method I was calling would be considering this extension.  So the reason my test is passing is the lack of a field type and the absence of pruning? 

> * logic to handle defaults under ObjectMeta correctly.

Special-casing presumably - not all fields can be defaulted?

Reply:
> I'm going to treat this as an educational opportunity. :)
> > note that this is by far not complete. This lacks
> > 
> > * the apiextensions/v1 validation
> 
> What is this and how is it relevant to default validation?

pkg/apiextensions/validation

It determines whether a default is allowed at all. It forbid certain schema fields although they show up in the schema types. For example we forbid defaults for fields in metadata.

> > > * a check for structurality of the provided schema
> 
> This method is validating a default value, not a schema. Am I missing something?

Only structural schemas allow defaults. But I agree: if you do the structural check elsewhere, we can probbaly assume it here.

> > > * a check that we actually have a v1 CRD
> As previously discussed, the current architecture of schema generation will likely require removal of generated defaults via post-processing.
> 
> > * validation that the value is pruned
> > Is this something I should just ignore then?

I think you can only ignore this requirement with a reasonable amount of code. It is tricky in combination with metadata and x-kubernetes-embedded-object. In other words, we have to be aware that the validation here is incomplete. This can be fine.

> > > * logic to cope with `x-kubernetes-preserve-unknown-fields`
> > Hmm, I was optimistically hoping that the validation method I was calling would be considering this extension. So the reason my test is passing is the lack of a field type and the absence of pruning?

I meant the pruning check here. The validation you call does not check pruning.

> 
> > * logic to handle defaults under ObjectMeta correctly.
> 
> Special-casing presumably - not all fields can be defaulted?

`<-- snip -->`

tl/dr: we can call validation via go-openapi here. That's fine and might catch errors from users. But it will be incomplete. I.e. you will be able to define APIs with markers which will result in CRD manifests which are rejected. Be clear about that in a Go comment at the top that this is best-effort.

After accepting that it is incomplete, we have to make sure the validation is correct, i.e. that you don't forbid defaults which would be valid in a CRD. I believe this is the case as we don't drop anything from the CRD schema, but only add more restriction through the `x-kubernetes-*` fields.